### PR TITLE
feat: Multichain Search DB export retry queue

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/11860

## Motivation

Currently, there is no mechanism to retry a failed initial import to the Multichain search DB.

## Changelog

Docs update https://github.com/blockscout/docs/pull/406.

### AI Agent summary

This pull request introduces significant changes to enhance the functionality and maintainability of the `Explorer` application. Key updates include the addition of a retry queue for failed database exports, refactoring of query methods, and improvements to helper modules for better code reuse. Below is a categorized summary of the most important changes:

### New Features
* Added the `Explorer.Chain.MultichainSearchDbExportRetryQueue` module to track data pending retry for export to the Multichain Service database. This includes a schema, changeset, and query methods for managing retryable operations.
* Introduced a migration to create the `multichain_search_db_export_retry_queue` table with `hash` and `hash_type` as primary keys and a new ENUM type for `hash_type`.

### Refactoring and Code Reuse
* Refactored transaction and block query methods to use a consistent `by_hashes_query` naming convention across modules (`Transaction`, `Block`, and `MultichainSearchDbExportRetryQueue`). [[1]](diffhunk://#diff-3fce0f082a5cb927aa738ef8f7d582824df91d08957971d36a0873efacd88816L1505-R1505) [[2]](diffhunk://#diff-068e2fe344a1d56ad5f54cda1ddcd99e420ee91f31673167307582418f9d3e24R583-R599) [[3]](diffhunk://#diff-75a1478214800d87c852cdcad18edae6e2abe1e15fb2fa261dc200eca1ae3d30L1366-R1367)
* Moved the `add_timestamps/1` function to a new `Explorer.Helper` module for better reusability and updated all references to use the helper. [[1]](diffhunk://#diff-18c8fc0f4efc97351c3c9310120ff34878e8b927eeecd8908c721d1145231253L7-R7) [[2]](diffhunk://#diff-18c8fc0f4efc97351c3c9310120ff34878e8b927eeecd8908c721d1145231253L71-R71) [[3]](diffhunk://#diff-18c8fc0f4efc97351c3c9310120ff34878e8b927eeecd8908c721d1145231253L104-R104) [[4]](diffhunk://#diff-4da038ce73822f037499c565936675b14a5295a317d25bf9539b448421aef73dR488-R510)

### Enhancements to Multichain Search Interface
* Extended the `batch_import/2` function in `Explorer.MicroserviceInterfaces.MultichainSearch` to handle retries for failed exports by inserting failed data into the retry queue. [[1]](diffhunk://#diff-4df7b4dbbb83cc7019726f6f9832ba5b00d4e8c02cd006a4ce0fab194b6175c9L34-R105) [[2]](diffhunk://#diff-4df7b4dbbb83cc7019726f6f9832ba5b00d4e8c02cd006a4ce0fab194b6175c9R126-R128)
* Added utility functions (`format_blocks`, `format_transactions`, etc.) to simplify the formatting of data for Multichain Service requests. [[1]](diffhunk://#diff-4df7b4dbbb83cc7019726f6f9832ba5b00d4e8c02cd006a4ce0fab194b6175c9R206-R251) [[2]](diffhunk://#diff-4df7b4dbbb83cc7019726f6f9832ba5b00d4e8c02cd006a4ce0fab194b6175c9L120-L158)

### Testing and Factory Updates
* Added a mock function `get_chain_id_mock/0` in `Explorer.TestHelper` for testing chain ID retrieval.
* Updated the factory module to include a factory for the `MultichainSearchDbExportRetryQueue` schema. [[1]](diffhunk://#diff-825aca4935bfc3fc143527f3ba20ccbdef83e5b5e20d1638c400e1a18dfa90e4R42) [[2]](diffhunk://#diff-825aca4935bfc3fc143527f3ba20ccbdef83e5b5e20d1638c400e1a18dfa90e4R748-R751)

These changes collectively improve the robustness of the application by introducing retry mechanisms, reducing code duplication, and enhancing testability.

#### Export retry functionality:
- init/3 initializes with data from the retry queue
Verifies that the init/3 function correctly initializes with data from the retry queue.

- run/2 successfully processes multichain search db export retry queue data
Ensures that run/2 processes retry queue data successfully when the external service responds with a 200 status.

- run/2 returns {:retry, export_data} on error
Confirms that run/2 returns {:retry, export_data} when the external service responds with an error (500 status).

Batch export functionality:
- batch_import/2 returns {:ok, :service_disabled} when the service is disabled
Verifies that batch_import/2 returns {:ok, :service_disabled} when the service is disabled.

- batch_import/2 processes chunks and returns {:ok, result} when the service is enabled
Ensures that batch_import/2 processes data chunks and returns {:ok, result} when the service is enabled.

- batch_import/2 returns {:error, reason} when an error occurs during processing and 'multichain_search_db_export_retry_queue' table is populated
Confirms that batch_import/2 returns {:error, reason} and populates the retry queue when an error occurs.

- batch_import/2 returns {:error, reason} when at least one chunk fails
Validates that batch_import/2 returns {:error, reason} when one or more chunks fail during processing.

- extract_batch_import_params_into_chunks/1 returns empty chunks when no data is provided
Validates that extract_batch_import_params_into_chunks/1 returns empty chunks when no input data is provided.

- extract_batch_import_params_into_chunks/1 returns chunks with transactions and blocks when no addresses are provided
Ensures that extract_batch_import_params_into_chunks/1 correctly handles transactions and blocks without addresses.

- extract_batch_import_params_into_chunks/1 returns chunks with correct structure for all data types
Verifies that extract_batch_import_params_into_chunks/1 returns properly structured chunks when all data types are provided.

- extract_batch_import_params_into_chunks/1 returns multiple chunks for large input data
Confirms that extract_batch_import_params_into_chunks/1 splits large input data into multiple chunks.

- extract_batch_import_params_into_chunks/1 returns chunks with correct structure for addresses only
Validates that extract_batch_import_params_into_chunks/1 handles input with only addresses correctly.

- extract_batch_import_params_into_chunks/1 returns chunks with correct structure of addresses
Ensures that extract_batch_import_params_into_chunks/1 includes detailed address information in the output chunks.

## Checklist for your Pull Request (PR)

- [x] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/setup/env-variables/README.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/setup/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/setup/env-variables).
- [x] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [x] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Summary by CodeRabbit

- **New Features**
  - Introduced a retry mechanism for exporting blockchain data to the Multichain Search DB, including a new retry queue and automated retry processing.
  - Added new functions and configuration options for improved error handling, retry queue management, and data chunking during batch exports.
  - Implemented a new database table for tracking failed export attempts with support for blocks, transactions, and addresses.
  - Enhanced batch import processing with concurrent chunk handling and enriched data formatting.
  - Added helper functions to manage timestamps and queue load during batch processing.
  - Integrated the retry fetcher into the supervision tree with configurable concurrency and batch size.
  - Improved asynchronous handling of batch imports to avoid blocking operations.

- **Bug Fixes**
  - Enhanced error handling to ensure failed exports are retried, logged, and queued appropriately.

- **Tests**
  - Added comprehensive tests for the retry logic, batch import functionality, data chunk extraction, and retry queue processing.

- **Chores**
  - Updated configuration files and environment variable placeholders to support new retry and batch processing features.

### CODE REVIEW

File: apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      )
+
+    acc
+  end
+
+  @impl BufferedTask
+  def run(export_data, _json_rpc_named_arguments) when is_list(export_data) do
+    pre_prepared_export_data =
+      export_data
+      |> Enum.reduce(%{address_hashes: [], block_hashes: [], transactions: []}, fn %{
```
Comment (fedor-ivn) [inline apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex:51]: Do I miss something or transactions list is unused below?

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 2]
```diff
@@ -1363,8 +1363,8 @@ defmodule Explorer.Chain.Transaction do
   @doc """
   Builds an `Ecto.Query` to fetch transactions by hashes
   """
-  @spec transactions_by_hashes([Hash.t()]) :: Ecto.Query.t()
-  def transactions_by_hashes(hashes) do
+  @spec transactions_by_hashes_query([Hash.t()]) :: Ecto.Query.t()
```
Comment (fedor-ivn) [inline apps/explorer/lib/explorer/chain/transaction.ex:1366]: very much imho nitpick: if we follow a shorter naming in other modules, can we rename this one as well (for consistency)?

```suggestion
  @spec by_hashes_query([Hash.t()]) :: Ecto.Query.t()
```

File: apps/indexer/lib/indexer/helper.ex [hunk 3]
```diff
@@ -883,4 +883,43 @@ defmodule Indexer.Helper do
     |> String.trim()
     |> String.trim_trailing("/")
   end
+
+  @max_queue_size 5000
```
Comment (nikitosing) [inline apps/indexer/lib/indexer/helper.ex:887]: Shouldn't we add possibility to configure queue size, at least for multichain importer?

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 4]
```diff
@@ -31,35 +33,81 @@ defmodule Explorer.MicroserviceInterfaces.MultichainSearch do
...
-        {:error, _reason}, _acc -> {:halt, {:error, @request_error_msg}}
+      |> Enum.reduce_while({:ok, {:chunks_processed, params_chunks}}, fn
+        {:ok, {:ok, _result}}, acc ->
+          {:cont, acc}
+
+        {:ok, {:error, _reason}}, _acc ->
+          {:halt, {:error, @request_error_msg}}
+
+        {:exit, _}, _acc ->
+          {:halt, {:error, @request_error_msg}}
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:62]: Try to set 
```elixir
@addresses_chunk_size 2
@max_concurrency 3
```
and run the following test
```
    test "returns {:error, reason} when an error occurs during processing and 'multichain_search_db_export_retry_queue' table is populated with all the input data" do
      bypass = Bypass.open()
      Application.put_env(:explorer, MultichainSearch, service_url: "http://localhost:#{bypass.port}", api_key: "12345")

      on_exit(fn ->
        Application.put_env(:explorer, MultichainSearch, service_url: nil, api_key: nil)
        Bypass.down(bypass)
      end)

      TestHelper.get_chain_id_mock()

      Bypass.expect(bypass, "POST", "/api/v1/import:batch", fn conn ->
        Conn.resp(
          conn,
          500,
          Jason.encode!(%{"code" => 0, "message" => "Error"})
        )
      end)

      assert Repo.aggregate(MultichainSearchDbExportRetryQueue, :count, :hash) == 0

      addresses = 10 |> insert_list(:address)

      params = %{
        addresses: addresses,
        blocks: [],
        transactions: []
      }

      assert {:error, @error_msg} = MultichainSearch.batch_import(params)

      assert Repo.aggregate(MultichainSearchDbExportRetryQueue, :count, :hash) == 10
    end
```

I faced strange unrelated errors, however once in like 5 attempts test fails with only 6 insertions to MultichainSearchDbExportRetryQueue

Probably, we need to create a separate issue, so the tests can work adequately with similar async code

File: apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        :ok
+
+      {:error, _} ->
+        Logger.error(fn ->
+          ["#{@failed_to_re_export_data_error}", "#{inspect(prepared_export_data)}"]
+        end)
+
+        {:retry, export_data}
+    end
+  end
```
Comment (sl1depengwyn) [inline apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex:117]: Maybe we can differentiate successful and unsuccessful imports so we do not retry same batch over and over? As far as I understand Multichain Search is quite heavy loaded, so it's better not to load it more

**Review by sl1depengwyn — CHANGES_REQUESTED** [review]
Also, maybe now it's possible to factor out multichain search import from main syncrous blocks import?

File: apps/indexer/lib/indexer/helper.ex (hunk 3 above)
Comment (vbaranov) [reply apps/indexer/lib/indexer/helper.ex:887]: Good suggestion - it will improve the manageability of the integration. Implemented in https://github.com/blockscout/blockscout/pull/12377/commits/54e974c001cb27dc64529e4d950c904136ee5ac4.

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 6]
```diff
@@ -31,35 +33,81 @@ defmodule Explorer.MicroserviceInterfaces.MultichainSearch do
...
+      do:
+        Repo.insert_all(MultichainSearchDbExportRetryQueue, Helper.add_timestamps(prepared_data), on_conflict: :nothing)
+  end
+
+  @spec http_post_request(String.t(), map(), boolean()) :: {:ok, any()} | {:error, String.t()}
+  defp http_post_request(url, body, retry?) do
     headers = [{"Content-Type", "application/json"}]
 
-    case HTTPoison.post(url, Jason.encode!(body), headers, recv_timeout: @post_timeout, hackney: [pool: false]) do
+    case (Application.get_env(:explorer, :http_adapter) || HTTPoison).post(url, Jason.encode!(body), headers,
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex]: Yes, it was added specifically for that case. I double-checked to ensure that we aren’t using a fallback in similar instances of the Application.get_env(:explorer, :http_adapter) call. For the sake of consistency, and considering that missing configuration is a very rare scenario - I’ve removed the fallback from this line in https://github.com/blockscout/blockscout/commit/54e974c001cb27dc64529e4d950c904136ee5ac4.

File: apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex (hunk 5 above)
Comment (vbaranov) [reply apps/indexer/lib/indexer/fetcher/multichain_search_db_export/retry.ex:117]: Good suggestion for future improvement—I've created an issue to track it: https://github.com/blockscout/blockscout/issues/12422.

That said, I don’t believe this is a blocker for the current PR. In the vast majority of retry cases where more than one chunk is involved, there will be only two chunks: the first (full) one, which fails, and a second, much smaller chunk, which is likely to succeed. Both are handled by the current retry logic. Retrying the second, smaller chunk shouldn't cause any significant issues.

**Comment by vbaranov** [comment]
> Also, maybe now it's possible to factor out multichain search import from main syncrous blocks import?

I don’t believe we’re ready to take that step at this stage, and I don’t see the current synchronous block import as a major issue for now. However, I’ve created an issue to track this as a potential future improvement: https://github.com/blockscout/blockscout/issues/12423.
Feel free to add your suggestions there on how you envision factoring this out.

**Comment by sl1depengwyn** [comment]
> I don’t believe we’re ready to take that step at this stage, and I don’t see the current synchronous block import as a major issue for now. However, I’ve created an issue to track this as a potential future improvement: https://github.com/blockscout/blockscout/issues/12423.
> Feel free to add your suggestions there on how you envision factoring this out.
I agree that this is out of scope of this pull request, however I think this is quite a serious issue as time of block import is crucial metric for indexer and now it's increased with with potentially quite big request to remote microservice that performs db insert operations and can be busy handling requests from other chains while we can wait for response up to 5 minutes, that is unacceptable time for,  for instance, new block being reflected on the chain

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex (hunk 4 above)
Comment (sl1depengwyn) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:61]: This test indicates a problem when data is being chunked:
For instance, let's take default 
```elixir
  @addresses_chunk_size 7_000
  @max_concurrency 5
```
Imagine calling
```elixir
batch_import(%{
            addresses: insert_list(35001, :address),
            blocks: [],
            transactions: []
          },
          false
        )
```
This call will create 6 chunks: `[7000], [7000], [7000], [7000], [7000], [1]`
5 first chunks will begin to import, however as soon as first of them errors, we halt reducing the task stream and return error, so we do not begin to process `[1]` and do not insert in the database to retry
I think this corner case should be handled

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 7]
```diff
@@ -145,17 +220,85 @@ defmodule Explorer.MicroserviceInterfaces.MultichainSearch do
...
+    else
+      Enum.map(indexed_addresses_chunks, fn {addresses_chunk, index} ->
+        # credo:disable-for-lines:2 Credo.Check.Refactor.Nesting
+        hashes_in_chunk = if index == 0, do: block_transaction_hashes, else: []
+        block_ranges_in_chunk = if index == 0, do: block_ranges, else: []
+
+        base_data_chunk
+        |> Map.put(:addresses, addresses_chunk)
+        |> Map.put(:hashes, hashes_in_chunk)
+        |> Map.put(:block_ranges, block_ranges_in_chunk)
```
[the remark this answers is not in the document]
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:247]: > maybe put hashes to last chunk? this way chunks would be more uniform, otherwise first chunk always has 7k addresses and all block and transactions hashes
not sure if it affects anything, but I suppose that if we divide data by chunks it's better to get more consistent chunks

In my opinion, the main challenge in the export process lies with addresses. Even when importing a single block with just a few transactions, the number of involved addresses can be extremely large, for example, in cases like airdrops. In my view, the “fat” chunks will be dominated by a high volume of addresses, with only a small number of block and transaction hashes included.

Inserting block and transaction hashes should be very fast. These are typically written once, so there’s no conflict handling involved, and the database insertion of this part should complete without delay.

With that in mind, I don’t believe separating block or transaction hashes into their own chunk would noticeably improve import speed. It doesn’t really matter which chunk they’re part of. What truly impacts performance is the size of the address chunks and how frequently addresses are imported, since that affects locking and potential deadlocks at the database level.

To partially address this, I introduced a runtime variable to control the maximum size of address chunks during export:
https://github.com/blockscout/blockscout/pull/12377/commits/0f550df2ff7c281515668bf5d48ce9131948f26e.

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex (hunk 4 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex]: You're right, this case wasn’t handled previously. I believe it’s now addressed in the following commit:
https://github.com/blockscout/blockscout/pull/12377/commits/f4253897c66354b827181062380b7f1061a9f2d7

Error handling during import has been moved to the synchronous part of the process. This ensures that all import chunks complete safely, even when failures occur, and helps prevent intermittent test failures.

I've also added two new tests based on the edge case you described:
- One that covers the scenario where all chunks fail.
- Another, that simulates the same behavior, but with the number of chunks exceeding the max_concurrency limit of Task.async_stream.

Additionally, I’ve moved the Multichain Search DB export into an asynchronous task during block data import. This change helps prevent slowdowns in block imports caused by issues in the Multichain Search API.

From what I can see, the only remaining weakness in the Multichain Search export is that it retries the entire batch if a single chunk fails—a limitation we've already agreed to address as a future enhancement.

Let me know if you spot any other issues or have further suggestions.

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex [hunk 8]
```diff
@@ -31,131 +32,311 @@ defmodule Explorer.MicroserviceInterfaces.MultichainSearch do
...
+          on_error(
+            %{
+              url: url,
+              data_to_retry: export_body,
+              reason: reason
+            },
+            retry?
+          )
+
+          {:cont, {:error, @request_error_msg}}
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex:73]: Nitpick: seems that now we don't neeed reduce_while, reduce is enough

File: apps/explorer/lib/explorer/chain.ex [hunk 9]
```diff
@@ -1169,10 +1169,11 @@ defmodule Explorer.Chain do
...
           # credo:disable-for-next-line Credo.Check.Refactor.Nesting
-          case MultichainSearch.batch_import(assets_to_import) do
-            {:ok, _} -> result
-            _ -> {:error, :insert_to_multichain_search_db_failed}
-          end
+          Task.async(fn ->
+            MultichainSearch.batch_import(assets_to_import)
+          end)
+
+          result
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain.ex:1176]: According to docs 
```
If you start an async, you must await. This is either done by calling Task.await/2 or Task.yield/2 followed by Task.shutdown/2 on the returned task. Alternatively, if you spawn a task inside a GenServer, then the GenServer will automatically await for you and call GenServer.handle_info/2 with the task response and associated :DOWN message.
```
We need to recheck that task in `Explorer.Chain.import` won't cause any problems, maybe we can remove import here and leave it in block fetchers (where it was removed)? Or this way we may loose some addresses from other fetchers?

File: apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex (hunk 8 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/microservice_interfaces/multichain_search.ex]: Yep, addressed in https://github.com/blockscout/blockscout/pull/12377/commits/622ce397a90257102ee32f1dac453fbbd0015baf.

File: apps/explorer/lib/explorer/chain.ex (hunk 9 above)
Comment (vbaranov) [reply apps/explorer/lib/explorer/chain.ex]: >  Or this way we may loose some addresses from other fetchers?

Yes, this change will result in the loss of address data export from secondary fetchers.
I suggest handling asynchronous batch export to multichain separately under the scope of [issue #12423](https://github.com/blockscout/blockscout/issues/12423), as I don’t currently see a quick and robust solution for it.
In the meantime, I’ve restored synchronous batch import in [this commit](https://github.com/blockscout/blockscout/pull/12377/commits/08d4db4f323ce28ad8a4748e80006562311156e0).

**Review by nikitosing — APPROVED** [review]

**Review by sl1depengwyn — APPROVED** [review]
